src/rgw/rgw_perf_counters.cc

```diff
@@ -3,69 +3,212 @@
+  lpcb->add_time_avg(l_rgw_op_get_lat, "get_initial_lat", "Get latency");
+
+  lpcb->add_u64_counter(l_rgw_op_del_obj, "del_obj_ops", "Delete objects");
+  lpcb->add_u64_counter(l_rgw_op_del_obj_b, "del_obj_bytes", "Size of delete objects");
+  lpcb->add_time_avg(l_rgw_op_del_obj_lat, "del_obj_lat", "Delete object latency");
+
+  lpcb->add_u64_counter(l_rgw_op_del_bucket, "del_bucket_ops", "Delete Buckets");
+  lpcb->add_time_avg(l_rgw_op_del_bucket_lat, "del_bucket_lat", "Delete bucket latency");
+
+  lpcb->add_u64_counter(l_rgw_op_copy_obj, "copy_obj_ops", "Copy objects");
+  lpcb->add_u64_counter(l_rgw_op_copy_obj_b, "copy_obj_bytes", "Size of copy objects");
+  lpcb->add_time_avg(l_rgw_op_copy_obj_lat, "copy_obj_lat", "Copy object latency");
+
+  lpcb->add_u64_counter(l_rgw_op_list_obj, "list_obj_ops", "List objects");
+  lpcb->add_time_avg(l_rgw_op_list_obj_lat, "list_obj_lat", "List objects latency");
+
+  lpcb->add_u64_counter(l_rgw_op_list_buckets, "list_buckets_ops", "List buckets");
+  lpcb->add_time_avg(l_rgw_op_list_buckets_lat, "list_buckets_lat", "List buckets latency");
+}
+
+std::shared_ptr<PerfCounters> create_rgw_counters(const std::string& name, CephContext *cct) {
+  std::string_view key = ceph::perf_counters::key_name(name);
+  if (rgw_op_counters_key.compare(key) == 0) {
+    PerfCountersBuilder pcb(cct, name, l_rgw_op_first, l_rgw_op_last);
+    add_rgw_op_counters(&pcb);
+    std::shared_ptr<PerfCounters> new_counters(pcb.create_perf_counters());
+    cct->get_perfcounters_collection()->add(new_counters.get());
+    return new_counters;
+  } else {
+    PerfCountersBuilder pcb(cct, name, l_rgw_first, l_rgw_last);
+    add_rgw_frontend_counters(&pcb);
+    std::shared_ptr<PerfCounters> new_counters(pcb.create_perf_counters());
+    cct->get_perfcounters_collection()->add(new_counters.get());
+    return new_counters;
+  }
+}
+
+void frontend_counters_init(CephContext *cct) {
+  PerfCountersBuilder pcb(cct, "rgw", l_rgw_first, l_rgw_last);
+  add_rgw_frontend_counters(&pcb);
+  PerfCounters *new_counters = pcb.create_perf_counters();
+  cct->get_perfcounters_collection()->add(new_counters);
+  perfcounter = new_counters;
+}
+
+namespace rgw::op_counters {
+
+PerfCounters *global_op_counters = NULL;
+
+void global_op_counters_init(CephContext *cct) {
+  PerfCountersBuilder pcb(cct, rgw_op_counters_key, l_rgw_op_first, l_rgw_op_last);
+  add_rgw_op_counters(&pcb);
+  PerfCounters *new_counters = pcb.create_perf_counters();
+  cct->get_perfcounters_collection()->add(new_counters);
+  global_op_counters = new_counters;
+}
+
+CountersContainer get(req_state *s) {
+  CountersContainer user_bucket_counters;
+
+  if (user_counters_cache && !s->user->get_id().id.empty()) {
+    std::string key = ceph::perf_counters::key_create(rgw_op_counters_key, {{"User", s->user->get_id().id}});
+    auto user_counters = user_counters_cache->get(key);
+    user_bucket_counters.user_counters = user_counters;
```

Comment:
nit: this variable naming is quite verbose. consider `counters.user`

and because `user_counters` is a shared_ptr, we don't want to construct it *and* copy it, because the copy requires an extra atomic ref and deref - just assign directly like:
```c++
counters.user = user_counters_cache->get(key);
```